cmd/docker-proxy/bin_windows_386.go

```diff
@@ -0,0 +1,8 @@
+//go:build windows && 386
+// +build windows,386
+
+//go:generate goversioninfo -o=../../cli/winresources/docker-proxy/resource.syso -icon=winresources/docker-proxy.ico -manifest=winresources/docker-proxy.exe.manifest ../../cli/winresources/docker-proxy/versioninfo.json
```

Comment:
Would it be better to have the output resource.syso be `resource_<platform>_<arch>.syso`? Then it could perhaps be committed if feasible, and it definitely wouldn't be at-risk of doing builds for different platforms/arches in a row and forgetting to `go generate` between each one.

I thought the usual idiom was to commit the `go generate`-created artefacts to the repo, but I'm working from a _really_ limited set of inputs there. But in this case, I think I understand some of the inputs are going to vary at build-time, and we have the Makefile-based system already to trigger `go generate` on-demand.

Reply:
I don't think resources should be committed but generated at build time. But yes we could document that `go generate` should be run before build even if it's already the case in our Makefile.

Reply:
It'd be nice if there was a way to make it _fail_ to build if `go generate` hadn't been run.

Reply:
> I thought the usual idiom was to commit the go generate-created artefacts to the repo

That's what I thought as well. I went looking for [a primary source](https://go.dev/blog/generate) and found instead: (emphasis mine)
> Just keep in mind that it is for package authors, not clients, if only for the reason that the program it invokes might not be available on the target machine. Also, **if the containing package is intended for import by `go get`**, once the file is generated (and tested!) it must be checked into the source code repository to be available to clients.

So in the specific case of generated files used only as inputs for producing binary artifacts, it is fair game to require that they be generated at build time. Though perhaps committing .syso files for "stub" Windows resources containing only the manifest (and no version/copyright info) into the repo might be worth considering? Then people developing & testing on Windows could use `go build` to compile manifested executables, and only those building for release (i.e. packaging maintainers) need to have goversioninfo installed. Stub .syso files could be committed and `.gitignore`'d so that non-stub versions are not accidentally committed; `git add --force` can be used to update the stubs when necessary.

> It'd be nice if there was a way to make it _fail_ to build if `go generate` hadn't been run. But nothing comes immediately to mind.

Commit some kind of "poisoned" deliberately corrupt .syso which causes link to fail unless overwritten with a generated version? I can't imagine the resulting build error messages would be terribly informative, however. I think that out of all the options, committing stub .syso files would be the least-terrible choice as that would ensure that the Windows binaries will always be correctly manifested and therefore behave correctly at runtime, even if the version info is missing.

Reply:
Yeah, I feel that a "poison" .syso is a worse-than-doing-nothing solution from a UX perspective.

I'm unsure about stub sysos. That's what hcsshim does for its test suite currently, but in that case there wasn't a go-native solution available, they were hand-created once using the same MSVC tools that you're replacing in the Docker pipeline here. And there's a TODO in the code to fix that anyway, although it wasn't mentioned if the fix is removing the syso and making it dynamic, or just having repeatable tooling to update the .syso when necessary.

I don't think this is a huge issue anyway, `go build` outside the Dockerfile isn't a super-supported workflow on Windows. I've used it on rare occasion, but I had to work out the autogen stuff for myself and hunt down various necessary binaries, so this is a big step forward compared to that in terms of clarify of tooling.

So I reckon what we have now here is fine.

As another (idle) thought, is `goversioninfo` runnable using `go run`? That's how one uses, e.g. [`mksyscall_windows`](https://github.com/microsoft/go-winio/blob/master/syscall.go), and then it gets built/installed automatically at `go generate` time, which means the person building their local dockerd.exe or unit tests doesn't need to separately deal with it.

`go run` might not work well with the build system in this project though, as it wouldn't be able to use the pre-built goversioninfo binary organised in a previous layer, AFAIK. I'd use `go run` if I was working in a simple Go-tooling-only build environment though, sometimes [through a wrapper if the tool itself is not `go run`-able](https://github.com/TBBle/hcsshim/commit/1803a218685ea1c449e1383639dbb459b1473fc7#diff-41f96ff92ea57c6d9e30f148d368d27f2a602885bf2acef714efb544d20ba668).